Impove Onboarding / Membership management process

markets opened the pull request:
Ref: https://github.com/coopdevs/timeoverflow/pull/770

Some improvements on Register & Membership process:


- [x] ~Review register process (create account and confirm email), some users get stuck in this process~ 
ℹ️ **NOTE** can't reproduce
- [x] Remember users without membership (15 days after register) they should ask for membership in some Organization or the user will be deleted when user have 1 month on the platform without membersehip
- [x] Super admin (active admin) allow to filter users without membership (we will manually manage this list and remove old users without membership)
- [x] Fix duplicate applications. Closes #763.
ℹ️ **NOTE** Added ability to remove non-accepted petitions from the admin 
- [x] Force page https://www.timeoverflow.org/organizations to users logged in without membership (some users don't know what to do if they see pages like https://www.timeoverflow.org/pages/about)
- [x] Improve text of email notifications new membership application, specify the name of the organization user want to join
- [x] Notify orgs admins (CCO) when membership remove their membership
- [x] On https://www.timeoverflow.org/organizations Organizations list place BdTOnline (org 402) always at the begining (it's the TimeBank that you can join if you don't find any in your location) 
ℹ️ **NOTE** Added `highlighted` flag to avoid hardcoding IDs

markets commented on app/views/organizations/index.html.erb:
```diff
@@ -31,7 +31,7 @@
       </thead>
       <tbody>
-        <%= render partial: 'organizations_row', collection: @user_organizations, as: :org if !params[:page] || params[:page] == '1' %>
+        <%= render partial: 'organizations_row', collection: @user_organizations, as: :org if @user_organizations.present? %>
```
**Self-note** Refactor this part

markets commented:
@sseerrggii I'm still polishing some details and working in some points, but I'm unable to reproduce the 1st issue 🤔: 

> Review register process (create account and confirm email), some users get stuck in this process

Do you have more specific information about those problems?

markets commented:
@franpb14 Do you want to take a quick look and give me your feedback?

franpb14 commented on app/controllers/members_controller.rb:
```diff
@@ -6,6 +6,8 @@ def destroy
     toggle_active_posts
     @member.destroy
 
+    OrganizationNotifier.member_deleted(@member).deliver_now
```
Just curious, why `deliver_now` instead of `deliver_later`?

markets replied:
Right! I just copy-pasted the same workflow we are already using for emails, but I think this is probably something from the past. I'll move all email delivery to the background queues.

markets replied:
Migrated to `deliver_later`.

franpb14 commented on app/mailers/organization_notifier.rb:
```diff
@@ -15,23 +15,37 @@ def recent_posts(posts, locale, users)
+    org_managers = organization.all_managers
 
-    I18n.with_locale(locale) do
+    I18n.with_locale(org_managers.first&.locale) do
```
if `org_managers.first` is nil will this work?

franpb14 commented on app/models/user.rb:
```diff
@@ -38,6 +38,7 @@ class User < ApplicationRecord
   accepts_nested_attributes_for :members, allow_destroy: true
 
   default_scope { order("users.id ASC") }
+  scope :without_memberships, -> { where.missing(:members) }
```
beautiful ✨

franpb14 commented on spec/controllers/statistics_controller_spec.rb:
```diff
@@ -45,7 +45,7 @@
       get :demographics
 
-      expect(assigns(:age_counts)).to eq({ "35-44" => 1 })
+      expect(assigns(:age_counts)).to eq({ "45-54" => 1 })
```
Maybe instead of harcoding this we could create the member using something like `date_of_birth: Time.now - 46.years`

sseerrggii commented:
> ll polishing some details and working in some points, but I'm unable to reproduce the 1

No problem, we will test whole process

Mariaadbdt commented:
Hello team! Thank you very much for the updates! I've been testing some of the features and I’m posting my feedback after each item of the list. 

- _Remember users without membership (15 days after register) they should ask for membership in some Organization or the user will be deleted when user have 1 month on the platform without membersehip._

Feedback: I registered as new user and the message reminding me to join a timebank appeared in 3 occasions. Before (https://staging.deploy.timeoverflow.org/members/please_confirm) and after (https://staging.deploy.timeoverflow.org/organizations) confirming the email. Then I logged out and in again and it also appeared upon requesting to join a timebank.
 
I would suggest keeping it only after the email confirmation and if possible send a reminder email within 15 days. And could you please review the text to make the process a bit clearer? Suggestion: Your registration will be completed once you have requested to join a timebank and the timebank accepts your request. If you do not request to join a timebank within 30 days, your user will be automatically deleted from the system.

- _Super admin (active admin) allow to filter users without membership (we will manually manage this list and remove old users without membership)_

Feedback: It works!

_- Fix duplicate applications. Closes_

Feedback/query: This is not yet in production, right? In Sants we still see the duplicates.

- _Force page https://www.timeoverflow.org/organizations to users logged in without membership (some users don't know what to do if they see pages like https://www.timeoverflow.org/pages/about)_

Feedback: It works both during the registration process and when I logged out and logged in again.

- _Improve text of email notifications new membership application, specify the name of the organization user want to join._ 

Feedback: I couldn’t test this, and I have a suggestion for the text: “Hello, XXX has just requested to join xxx Timebank. You can enter …. to accept or reject the request. Once you accept the request the user will automatically appear on your users list.”

- _Notify_ orgs admins (CCO) when membership remove their membership

Feedback: I couldn’t test this but it’s a great feature!

- _On_ https://www.timeoverflow.org/organizations Organizations list place BdTOnline (org 402) always at the begining (it's the TimeBank that you can join if you don't find any in your location) ℹ️ NOTE Added highlighted flag to avoid hardcoding IDs

Feedback: ok. 

Bonus track:
I accidently landed on this page (https://staging.deploy.timeoverflow.org/select_organization) after having deactivated the membership of my new user from the admin panel. In fact, I thought I had deleted the user and restarted the registration process of a new “new account”. I moved from the registration form to this error page. The sentence in the blue square needs revision as well, as it doesn’t give any “further info”. And I’m missing a “next”, “exit”, “continue”, “back” or whatever button to go to another page or suggest a next step.

![tof error](https://github.com/user-attachments/assets/aea43b45-840e-4eb6-a7d0-4acb9a032f49)

markets commented:
@Mariaadbdt thank you for your feedback! I'm going to address those points in the following days.

sseerrggii commented:
Thanks @Mariaadbdt for testing and feedback

This features are not in production, when we approve all the changes we will creat a new release of TimeOverflow on production

About the suggestion _"If you do not request to join a timebank within 30 days, your user will be automatically deleted from the system."_ In order to no add complexity to this development I will keep this work in manual work of ADBdT (super admin)

markets commented:
@Mariaadbdt @sseerrggii Note about this point:

> Fix duplicate applications

The new code avoids the creation of new duplicate applications, and in the other hand I introduced the ability to remove applications from the admin side. So even when we release this to production the old duplicated applications will be there and someone will need to do the clean up manually from the admin side.

Mariaadbdt commented:
Thank you for the prompt responses, guys!
@sseerrggii  you can take out the word "automatic" from the message and just inform that the user will be deleted. I'll add a reminder in my calendar every month to review and delete the unassigned users ;-)

markets commented:
@Mariaadbdt @sseerrggii I think I addressed all comments (I also added this "I would suggest keeping it only after the email confirmation and if possible send a reminder email within 15 days") and redeployed to staging. Could you please take a look and confirm?

I also think we can go with the manual deletion for now, but it can be easily automated in future releases (it will be mostly the same code as the new warning email).

franpb14 reviewed (approved):
the code looks good to me

sseerrggii commented:
* Notify orgs admins (CCO) when membership remove their membership

I don't see this email notification working, 

1. I register a new user 
2. request membership for two organization
3. Approve both requests with admin user
4. Relogin with the user and remove membership from one organization
5. Check on transactional mails platform logs and don't see any email notification

For the rest looks good to me. Great job @markets !!

markets commented:
Thanks for the feedback @sseerrggii I moved all the email delivery to our background jobs and it seems there is an issue with that refactor. I'll review it and ping you back for a final 👀.

markets commented:
@sseerrggii I think this point "Notify orgs admins (CCO) when membership remove their membership" is fixed now, could you please confirm 🙏🏼?

markets commented:
**UPDATE** Confirmed by @sseerrggii ✅ I'm going to merge it and prepare a new release in the following days.
